harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -268,75 +541,82 @@ def cleanup_iterator() -> None:
-                            "determined_version": det.__version__,
-                            "steps_completed": self.steps_completed,
-                            "framework": f"torch-{torch.__version__}",
-                            "format": "pickle",
-                        }
-                        with self.context._core.checkpoint.store_path(metadata) as (
-                            path,
-                            storage_id,
-                        ):
-                            self._save(path)
-                            uuid = storage_id
-                        response = {"uuid": storage_id}
-                    else:
-                        response = {}
-                    uuid = self.context.distributed.broadcast(uuid)
-                    for callback in self.callbacks.values():
-                        callback.on_checkpoint_upload_end(uuid=uuid)
+        return train_unit
 
-                else:
-                    raise AssertionError("Unexpected workload: {}".format(w.kind))
+    def _run(self):
+        min_checkpoint_period = self.env.experiment_config.get_min_checkpoint_period()
+        min_validation_period = self.env.experiment_config.get_min_validation_period()
 
-            except det.InvalidHP as e:
-                logging.info(f"Invalid hyperparameter exception during {action}: {e}")
-                response = workload.InvalidHP()
-            response_func(response)
-            self.upload_tb_files()
+        for op in self.core_context.searcher.operations():
+            self.searcher_op = op
+
+            train_unit = self._get_train_unit_from_op(op)
+
+            min_checkpoint_period = self.as_batches(**min_checkpoint_period)
+            min_validation_period = self.as_batches(**min_validation_period)
+            if min_checkpoint_period < 1:
+                min_checkpoint_period = sys.maxsize
+
+            if min_validation_period < 1:
+                min_validation_period = sys.maxsize
+
+            self._train(train_unit=train_unit,
+                        min_validation_period=min_validation_period,
+                        min_checkpoint_period=min_checkpoint_period)
+
+            logging.debug(f"Training finished after {train_unit.value} steps.")
+
+            # Report metrics after searcher complete if there are remaining unreported
+            if self.training_metrics:
+                self.report_training_metrics(self.steps_completed)
+
+            # Validate and report validation metrics if last validation step is not current step
+            if self.last_validation != self.steps_completed:
+                val_metrics = self._validate()
+                self.report_validation_metrics(self.steps_completed, val_metrics)
+
+            # Checkpoint
+            self.checkpoint(already_exiting=False)
+
+    def _validate_searcher_metric(self, val_metrics):
+        searcher_metric_name = self.env.experiment_config["searcher"]["metric"]
```

Comment:
another experiment-config-centric thing.

How should this look in a Trainer API, especially in local-training mode?